src/nvim/terminal.c

```diff
@@ -0,0 +1,744 @@
+        .rgb_fg_color = vt_fg != default_vt_fg ? vt_fg : -1,
+        .rgb_bg_color = vt_bg != default_vt_bg ? vt_bg : -1,
+      });
+    }
+
+    if (term->cursor.visible && term->cursor.row == row
+        && term->cursor.col == col) {
+      attr_id = hl_combine_attr(attr_id, get_attr_entry(&(attrentry_T) {
+        .rgb_ae_attr = 0, .rgb_fg_color = -1, .rgb_bg_color = -1,
+        .cterm_ae_attr = 0, .cterm_fg_color = 0, .cterm_bg_color = 11,
+      }));
+    } 
+
+    term_attrs[col] = attr_id;
+  }
+}
+
+static int term_damage(VTermRect rect, void *data)
+{
+  bool is_empty = true;
+  void *stub; (void)(stub);
+  map_foreach_value(damaged_terminals, stub, { is_empty = false; });
+
+  Terminal *term = data;
+  term->invalid_start = MIN(term->invalid_start, rect.start_row);
+  term->invalid_end = MAX(term->invalid_end, rect.end_row);
+  pmap_put(ptr_t)(damaged_terminals, data, NULL);
+
+  if (is_empty) {
+    event_push((Event) { .handler = on_damage }, false);
+  }
+
+  return 1;
+}
+
+static int term_moverect(VTermRect dest, VTermRect src, void *data)
+{
+  VTermRect rect_union = {
+    .start_row = MIN(dest.start_row, src.start_row),
+    .end_row = MAX(dest.end_row, src.end_row),
+  };
+  return term_damage(rect_union, data);
+}
+
+static int term_movecursor(VTermPos new, VTermPos old, int visible,
+    void *data)
+{
+  Terminal *term = data;
+  term->cursor.row = new.row;
+  term->cursor.col = new.col;
+
+  VTermRect old_rect = { .start_row = old.row, .end_row = old.row + 1 };
+  VTermRect new_rect = { .start_row = new.row, .end_row = new.row + 1 };
+  term_damage(old_rect, data);
+  term_damage(new_rect, data);
+  return 1;
+}
+
+static int term_settermprop(VTermProp prop, VTermValue *val, void *data)
+{
+  Terminal *term = data;
+
+  switch(prop) {
+    case VTERM_PROP_ALTSCREEN:
+      term->altscreen = val->boolean;
+      break;
+
+    case VTERM_PROP_CURSORVISIBLE:
+      term->cursor.visible = val->boolean;
+      break;
+
+    case VTERM_PROP_TITLE:
+      terminal_set_title(term, val->string);
+      redraw_buf_later(term->buf, CLEAR);
+      break;
+
+    case VTERM_PROP_MOUSE:
+      term->mouse_enabled = (bool)val->number;
+      break;
+
+    default:
+      return 0;
+  }
+
+  return 1;
+}
+
+static int term_bell(void *data)
+{
+  ui_putc('\x07');
+  return 1;
+}
+
+// the scrollback push/pop handlers were copied almost verbatim from pangoterm
+static int term_sb_push(int cols, const VTermScreenCell *cells, void *data)
+{
+  Terminal *term = data;
+  // save our buffer
+  buf_T *save_curbuf = NULL;
+  win_T *save_curwin = NULL;
+  tabpage_T *save_curtab = NULL;
+  switch_to_win_for_buf(term->buf, &save_curwin, &save_curtab, &save_curbuf);
+  // shift up the first visible line, it will be hidden and vterm will redraw
+  // over it
+  linenr_T src_linenr = (linenr_T)term->sb_current + 1;
+  linenr_T tgt_linenr = (linenr_T)term->sb_current;
+  ml_append(tgt_linenr, ml_get(src_linenr), 0, false);
```

Comment:
`ml_append()` description says `line` argument ( = `ml_get(src_linenr)`) should not be from another buffer. Is this safe?